### PR TITLE
[LLVMGPU] Improve reduction flow and support handled fused broadcast

### PR DESCRIPTION
**Pull request by ThomasRaoux**
Move buferization to after vectorization for group reduction pipeline.
This will allow better register level fusion. Also improve distribution
interaction with hal operations and add a test to make sure this can
handle fusion with broadcast.

### CODE REVIEW

**Comment by ThomasRaoux** [comment]
Depends on https://reviews.llvm.org/D130135 to pass lit tests

**Review by MaheshRavishankar — APPROVED** [review]
Nice!

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp [hunk 1]
```diff
@@ -184,11 +184,13 @@ void addGPUMatmulTensorCorePassPipeline(OpPassManager &pm,
...
 void addGPUWarpReductionPassPipeline(OpPassManager &pm) {
-  tileAndBufferize(pm);
-
-  // Don't tile to warp or thread and use vector distribution instead.
-
+  pm.addPass(createTileAndDistributeToWorkgroupsPass());
   auto &nestedModulePM = pm.nest<ModuleOp>();
+  nestedModulePM.addNestedPass<func::FuncOp>(
+      createConvertToDestinationPassingStylePass());
+  nestedModulePM.addPass(createCanonicalizerPass());
```
Comment (okkwon) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp:191]: We have back-to-back (canonicalizer + CSE)s. Is it intended?

File: compiler/src/iree/compiler/Codegen/LLVMGPU/VectorReductionToGPU.cpp [hunk 2]
```diff
@@ -82,23 +83,40 @@ static Value groupReduction(Location loc, OpBuilder &builder, Value input,
   return laneVal;
 }
 
-/// Special case to hoist hal operations that have side effect but are safe to
-/// move out of the warp single lane region.
-static void hoistHalBindingOps(vector::WarpExecuteOnLane0Op warpOp) {
+/// Hoist uniform operations as well as special hal operations that have side
+/// effect but are safe to move out of the warp single lane region.
+void static moveScalarAndBindingUniformCode(
```
Comment (okkwon) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/VectorReductionToGPU.cpp:88]: TIL `void static` is a thing. Interesting.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp (hunk 1 above)
Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp:191]: ah no, not sure how that happened

File: compiler/src/iree/compiler/Codegen/LLVMGPU/VectorReductionToGPU.cpp (hunk 2 above)
Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/VectorReductionToGPU.cpp]: oops... changing

**Review by okkwon — APPROVED** [review]
LGTM. Thanks for the change!

**Pull request merged**